core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2023 Amazon.com Inc. or its affiliates. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#ifndef __wasi__
+#error This example only compiles to WASM/WASI target
+#endif
+
+#include <assert.h>
+#include <errno.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <unistd.h>
+
+enum CONSTANTS {
+    NUM_ITER = 90000,
+    NUM_RETRY = 5,
+    MAX_NUM_THREADS = 8,
+    RETRY_SLEEP_TIME = 8000,
+};
+
+unsigned int threads_executed = 0;
+unsigned int threads_creation_tried = 0;
+
+void *
+thread_func(void *arg)
+{
+    (void)(arg);
+    __atomic_fetch_add(&threads_executed, 1, __ATOMIC_RELAXED);
+    return NULL;
+}
+
+void
+spawn_thread(pthread_t *thread)
+{
+    int status_code = -1;
+    for (int tries = 0; status_code != 0 && tries < NUM_RETRY; ++tries) {
+        status_code = pthread_create(thread, NULL, &thread_func, NULL);
+        ++threads_creation_tried;
+        assert(status_code == 0 || status_code == EAGAIN);
+        if (status_code == EAGAIN) {
+            usleep(RETRY_SLEEP_TIME);
+        }
+    }
+
+    assert(status_code == 0 && "Thread creation should succeed");
+}
+
+int
+main(int argc, char **argv)
+{
+    pthread_t threads[MAX_NUM_THREADS];
+    double percentage = 0.1;
+
+    for (int iter = 0; iter < NUM_ITER; ++iter) {
+        if (iter > NUM_ITER * percentage) {
+            fprintf(stderr, "Spawning stress test is %d%% finished\n",
+                    (unsigned int)(percentage * 100));
+            percentage += 0.1;
+        }
+
+        unsigned int thread_num = iter % MAX_NUM_THREADS;
+        if (threads[thread_num] != 0) {
```

Comment:
For this test I think we should spawn threads as quickly as possible - here we wait for a specific thread ID to be released, even though other IDs might have been freed already and could be used. I think this was discussed for the other test too, but I don't know if that was eventually addressed? If so, the code probably could be reused.

Reply:
I'd expect that we use this part of pthread_join

```  
If that thread has already terminated, then
       pthread_join() returns immediately. 
```

I'm curious whether there's any better solution. If we have mutex, thread function is going to be much longer I suppose. 

If we have atomic, we most likely need SEQ order, also not the fastest solution. 

I think that `thread_execution_time` is going to be less than `7 * thread_creation_time` for this tiny thread function. 

But it's just my assumation, tell me if I'm wrong

Reply:
So imagine we have three threads (call it 1, 2 and 3). Potential sequence can be:
```
Spawn thread 1
Spawn thread 2
Spawn thread 3
Finish thread 2
Finish thread 3
Spawn thread 1 <- block here
```
Because the first thread is still alive (e.g. due to scheduling), we can't spawn new thread even though slot 2 and 3 are freed already, so we wait unnecessarily.
Not sure how likely or severe the issue is, but it feels like we can completely remove the problem with a low overhead by not relying on the order.

Reply:
Fixed